src/main.rs

```diff
@@ -47,125 +97,293 @@ impl LogLevel {
-            }
-        },
-    }
-    let mut failure = false;
-    for path in [
-        "/usr/lib/greenboot/check/required.d/*.sh",
-        "/etc/greenboot/check/required.d/*.sh",
-    ] {
-        for entry in glob(path)?.flatten() {
-            log::info!("running required check {}", entry.to_string_lossy());
-            let output = Command::new("bash").arg("-C").arg(entry).output()?;
-            if !output.status.success() {
-                // combine and print stderr/stdout
-                log::warn!("required script failed...");
-                failure = true;
+        path_exists = true;
+        if let Some(errors) = run_scripts("required", &greenboot_required_path).err() {
+            log::error!("required script error:");
+            errors.iter().for_each(|e| log::error!("{e}"));
+            if !required_script_failure {
+                required_script_failure = true;
             }
         }
     }
-    for path in [
-        "/usr/lib/greenboot/check/wanted.d/*.sh",
-        "/etc/greenboot/check/wanted.d/*.sh",
-    ] {
-        for entry in glob(path)?.flatten() {
-            log::info!("running required check {}", entry.to_string_lossy());
-            let output = Command::new("bash").arg("-C").arg(entry).output()?;
-            if !output.status.success() {
-                // combine and print stderr/stdout
-                log::warn!("wanted script failed...");
-            }
+    if !path_exists {
+        bail!("required.d not found");
+    }
+    for path in GREENBOOT_INSTALL_PATHS {
+        let greenboot_wanted_path = format!("{path}/check/wanted.d/");
+        if let Some(errors) = run_scripts("wanted", &greenboot_wanted_path).err() {
+            log::warn!("wanted script runner error:");
+            errors.iter().for_each(|e| log::error!("{e}"));
         }
     }
-    // if a command with restart option in systemd fails to start we don't get it as "failed"
-    // reversing the check makes sure that if by the time After=multi-user the service isn't running then it's failing at least
-    let output = Command::new("systemctl")
-        .arg("list-units")
-        .arg("--state")
-        .arg("active")
-        .arg("--no-page")
-        .arg("--output")
-        .arg("json")
-        .output()?;
-    let services: Vec<ServiceStatus> = serde_json::from_str(&String::from_utf8(output.stdout)?)?;
-    let ss: Vec<String> = services.iter().map(|x| x.unit.clone()).collect();
-    let active_units: HashSet<String> = HashSet::from_iter(ss);
-    for service in ["sshd.service", "NetworkManager.service"] {
-        if !active_units.contains(service) {
-            log::warn!("service {} failed, see journal", service);
-            failure = true;
+
+    if required_script_failure {
+        bail!("health-check failed!");
+    }
+    Ok(())
+}
+
+/// runs all the scripts in red.d when health-check fails
+fn run_red() -> Result<(), Vec<Box<dyn Error>>> {
```

Comment:
we are not propagating any errors here either, and the function does not error, we should just return the vector with the collected errors

Reply:
we are propagating the err here:

```
run_red().unwrap_or_else(|errors| {
                log::error!("There is a problem with red script runner");
                errors.iter().for_each(|e| log::error!("{e}"));
            });
```

Reply:
but I get it we can only return vector as there is no result.

Reply:
that is not propagating, is handling the result, which is a vector of things that we log, we don't return the error